### DISCUSSION TITLE
redirect fails in Firewall\ContextListener if User has a File anywhere

### DISCUSSION THREAD

**Discussion by tacman** [symfony/symfony#54048] [Q&A]
How can I set the serialization groups on the context for an email?

I'm unable to serialize a File, and don't need to, but there's a file on an entity that I'm passing to an email template.  If I were simply dispatching a message, I could add the SerializerStamp, but I'm just following the mailer instructions and calling ->send().

I could manually serialize it before passing it, but wonder if there was a way to add stamps or something like that.

**Comment by MatTheCat** [comment]
You’d better solve the problem you have rather than its consequences: useless data don’t need to be sent so they don’t need to be serialized.

Create a dedicated object with only the data your email template needs; this will be simpler, more flexible, and more performant. Don’t limit yourself to your entities; you’re free to create as many classes as you need.

**Comment by tacman** [comment]
Thanks, I did that, now I realize that it's not just serializing for email but also for the Symfony Form redirect.

https://stackoverflow.com/questions/49782167/serialization-of-symfony-component-httpfoundation-file-file-is-not-allowed-sy

I have this exact situation -- I'm using VichUploader so users can submit a photo and some metadata (credit, place, etc.).  When the user uploads the photo, the Submission entity, which has a File object, is added to the user record ($user->addSubmission($submission), then persisted, and send the email to an admin to approve the photo before posting.  Using a DTO for the Submission record works fine.

But the final redirect triggers an errror.

            $redirect =  $this->redirectToRoute('submission_show', ['id' => $submission->getId()]);
            return $redirect;

This only fails when the submission has a user. And it doesn't have anything to do with the email anymore, so I've changed the title.  Submission has a User, a User has Submissions, the Submissions have a File.  I don't understand what the ContextListener has to do with redirecting after a form submission.  The form itself is valid.  dd($redirect) show what looks like a normal redirect.  

![image](https://github.com/symfony/symfony/assets/619585/91c6fd52-cb76-4911-bc47-4e82d6726f52)

**Comment by tacman** [comment]
A solution, which I don't understand, is to refresh the user before doing the redirect.

        if ($form->isSubmitted() && $form->isValid()) {
            if ($user) {
                $user->addSubmission($submission);
            }
            $entityManager = $this->entityManager;
            $entityManager->persist($submission);
            $entityManager->flush();
            $entityManager->refresh($user); 

            return $this->redirectToRoute(...

**Reply by MatTheCat** [reply]
Please use “Write a reply” instead of “Suggest an answer” in order to create legible conversations. I’ll answer here because this looks like a follow-up of https://github.com/symfony/symfony/discussions/54048#discussioncomment-8577318

The VichUploaderBundle has the exact same problem you were having: it stores data in unrelated objects. There is no reason you would want an instance of `File` in your entity.

The good news is that the (good) solution is the same as before: introduce an intermediary object as the form model, and then you’re free to use it as you want to update your entity (see https://verraes.net/2013/04/decoupling-symfony2-forms-from-entities/ e.g.).

FYI the bundle has a [`inject_on_load`](https://github.com/dustin10/VichUploaderBundle/blob/master/docs/usage.md#step-3-configure-the-lifecycle-events-optional-step) setting defaulting to `false`; this is why refreshing the user unsets the file and prevent the serialization issue.

(Also you don’t need to `persist` managed entities, unless your tracking policy is `DEFERRED_EXPLICIT` — which is probably not the case.)
